src/platform/nrfconnect/OTAMultiImageDownloader.cpp

```diff
@@ -0,0 +1,181 @@
+/*
+ *
+ *    Copyright (c) 2022 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "OTAMultiImageDownloader.h"
+#include <dfu/dfu_target.h>
+#include <platform/CHIPDeviceLayer.h>
+#include <pm_config.h>
+
+namespace chip {
+namespace DeviceLayer {
+
+size_t OTAMultiImageDownloaders::sCurrentId;
+
+std::array<OTAMultiImageDownloader, OTAMultiImageDownloaders::CurrentImageID::LIMIT> OTAMultiImageDownloaders::sImageHandlers{
+    { OTAMultiImageDownloader{ 0, PM_APP_1_CORE_APP_DOWNLOAD_OFFSET, PM_APP_1_CORE_APP_DOWNLOAD_SIZE, PM_APP_1_CORE_APP_ADDRESS,
+                               PM_APP_1_CORE_APP_SIZE },
+      OTAMultiImageDownloader{ 1, PM_APP_1_CORE_NET_DOWNLOAD_OFFSET, PM_APP_1_CORE_NET_DOWNLOAD_SIZE, PM_APP_1_CORE_NET_ADDRESS,
+                               PM_APP_1_CORE_NET_SIZE },
+      OTAMultiImageDownloader{ 2, PM_APP_2_CORE_APP_DOWNLOAD_OFFSET, PM_APP_2_CORE_APP_DOWNLOAD_SIZE, PM_APP_2_CORE_APP_ADDRESS,
+                               PM_APP_2_CORE_APP_SIZE },
+      OTAMultiImageDownloader{ 3, PM_APP_2_CORE_NET_DOWNLOAD_OFFSET, PM_APP_2_CORE_NET_DOWNLOAD_SIZE, PM_APP_2_CORE_NET_ADDRESS,
+                               PM_APP_2_CORE_NET_SIZE } }
+};
+
+static const device * sExtFlash = DEVICE_DT_GET(DT_CHOSEN(nordic_pm_ext_flash));
+
+int OTAMultiImageDownloader::Init(const device * flash, ssize_t size)
+{
+    if (!device_is_ready(flash))
+    {
+        ChipLogError(SoftwareUpdate, "Flash device is not ready!");
+        return -EFAULT;
+    }
+    return Init(flash, mOffset, size);
+}
+
+int OTAMultiImageDownloader::Init(const device * flash, size_t offset, size_t size)
+{
+    mBuffer = (uint8_t *) k_malloc(kBufferSize);
```

Comment:
I think using kernel heap may not be the best choice given how precious resource there is. We should better use `Platform::MemoryAlloc` or even `ScopedBuffer` for safer memory management.

Reply:
Ok, will do.